webapp/src/Controller/Team/SubmissionController.php

```diff
@@ -60,34 +61,111 @@ public function createAction(Request $request, ?Problem $problem = null): Respon
         }
-        $form    = $this->formFactory
+        $formUpload = $this->formFactory
             ->createBuilder(SubmitProblemType::class, $data)
             ->setAction($this->generateUrl('team_submit'))
             ->getForm();
 
-        $form->handleRequest($request);
+        $formPaste = $this->formFactory
+            ->createBuilder(SubmitProblemPasteType::class, $data)
+            ->setAction($this->generateUrl('team_submit'))
+            ->getForm();
 
-        if ($form->isSubmitted() && $form->isValid()) {
+        $formUpload->handleRequest($request);
+        $formPaste->handleRequest($request);
+
+        if ($formUpload->isSubmitted() && $formUpload->isValid()) {
             if ($contest === null) {
                 $this->addFlash('danger', 'No active contest');
             } elseif (!$this->dj->checkrole('jury') && !$contest->getFreezeData()->started()) {
                 $this->addFlash('danger', 'Contest has not yet started');
             } else {
                 /** @var Problem $problem */
-                $problem = $form->get('problem')->getData();
+                $problem = $formUpload->get('problem')->getData();
                 /** @var Language $language */
-                $language = $form->get('language')->getData();
+                $language = $formUpload->get('language')->getData();
                 /** @var UploadedFile[] $files */
-                $files      = $form->get('code')->getData();
+                $files      = $formUpload->get('code')->getData();
                 if (!is_array($files)) {
                     $files = [$files];
                 }
-                $entryPoint = $form->get('entry_point')->getData() ?: null;
+                $entryPoint = $formUpload->get('entry_point')->getData() ?: null;
                 $submission = $this->submissionService->submitSolution(
-                    $team, $this->dj->getUser(), $problem->getProbid(), $contest, $language, $files, 'team page', null,
-                    null, $entryPoint, null, null, $message
+                    $team,
+                    $this->dj->getUser(),
+                    $problem->getProbid(),
+                    $contest,
+                    $language,
+                    $files,
+                    'team page',
+                    null,
+                    null,
+                    $entryPoint,
+                    null,
+                    null,
+                    $message
+                );
+
+                if ($submission) {
+                    $this->addFlash(
+                        'success',
+                        'Submission done! Watch for the verdict in the list below.'
+                    );
+                } else {
+                    $this->addFlash('danger', $message);
+                }
+                return $this->redirectToRoute('team_index');
+            }
+        } elseif ($formPaste->isSubmitted() && $formPaste->isValid()) {
+            if ($contest === null) {
```

Comment:
A lot of those now get duplicated so maybe merge those where possible? Either by a helper function or using something like:
`if ($formUpload->isSubmitted() || $formPaste->isSubmitted())` and do the `if/elseif` one branch deeper.

Reply:
I think I've made it much clearer. Is this okay now?